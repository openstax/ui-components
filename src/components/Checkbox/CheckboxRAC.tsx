import styled from "styled-components";
import { colors } from "../../theme";
import {
  Checkbox as RACCheckbox,
  CheckboxProps as RACCheckboxProps,
  Text
} from "react-aria-components";
import { PropsWithChildren } from "react";
import { checkboxVariants, CheckboxSize, CheckboxVariant } from "./sharedCheckboxStyles";

export interface CheckboxRACProps
  extends PropsWithChildren<Omit<RACCheckboxProps, "children">> {
  size?: CheckboxSize;
  variant?: CheckboxVariant;
  bold?: boolean;
}

const StyledCheckbox = styled(RACCheckbox)<{
  $variant: CheckboxVariant;
  $size: CheckboxSize;
  $bold: boolean;
}>`
  display: flex;
  align-items: center;
  font-size: 1.6rem;
  font-weight: ${({ $bold }) => ($bold ? 700 : 400)};
  gap: 1.2rem;
  color: inherit;

  &[data-disabled] {
    opacity: 0.4;
    cursor: not-allowed;
  }

  [data-slot="selection"] {
    appearance: none;
    width: ${({ $size }) => $size}rem;
    height: ${({ $size }) => $size}rem;
    margin: 0;
    border-radius: 0.2rem;
    display: grid;
    place-content: center;
    border: ${({ $variant }) => checkboxVariants[$variant].unCheckedBorder};
    background-color: ${colors.palette.white};
    position: relative;

    &::before {
      content: "";
      position: absolute;
      inset: 0;
      border-radius: 0.2rem;
      background-color: ${({ $variant }) => checkboxVariants[$variant].backgroundColor};
      background-image: url('${({ $variant }) => checkboxVariants[$variant].backgroundImage}');
      background-size: 80%;
      background-position: center;
      background-repeat: no-repeat;
      transform: scale(0);
    }
  }

  &[data-selected] [data-slot="selection"]::before {
    transform: scale(1);
    opacity: ${(props => props.isDisabled ? 0 : 1)};
  }
`;

export const CheckboxRAC = ({
  size = 1.6,
  variant = "primary",
  bold = false,
  children,
  ...props
}: CheckboxRACProps) => {
  return (
    <StyledCheckbox
      {...props}
      $variant={variant}
      $size={size}
      $bold={bold}
    >
      <div slot="selection" data-slot="selection" />
      <Text>{children}</Text>
    </StyledCheckbox>
  );
};