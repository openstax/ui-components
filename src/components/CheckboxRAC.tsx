import { Checkbox as RACCheckbox, CheckboxProps as RACCheckboxProps } from "react-aria-components";
import styled from "styled-components";
import { PropsWithChildren } from "react";
import { whiteCheckmark, grayCheckmark, redCheckmark } from "./svgs/checkmarksvgs";
import { colors } from "../theme";

const checkboxVariants = {
  primary: {
    backgroundColor: colors.palette.mediumBlue,
    borderColor: colors.palette.mediumBlue,
    checkmark: whiteCheckmark
  },
  light: {
    backgroundColor: colors.palette.white,
    borderColor: colors.palette.pale,
    checkmark: grayCheckmark
  },
  error: {
    backgroundColor: colors.palette.paleRed,
    borderColor: colors.palette.lightRed,
    checkmark: redCheckmark
  },
  disabled: {
    backgroundColor: colors.palette.white,
    borderColor: colors.palette.pale,
    checkmark: 'none'
  }
} as const;

type Variant = keyof typeof checkboxVariants;
type Size = 1.4 | 1.6 | 1.8 | 2;

const StyledCheckbox = styled(RACCheckbox)<{
  $variant: Variant;
  $size: Size;
  $bold: boolean;
}>`
  display: flex;
  align-items: center;
  font-size: 1.6rem;
  font-weight: ${({ $bold }) => ($bold ? 700 : 400)};
  gap: 1.2rem;
  color: inherit;

  input[type="checkbox"] {
    appearance: none;
    width: ${({ $size }) => $size}rem;
    height: ${({ $size }) => $size}rem;
    margin: 0;
    display: grid;
    place-content: center;
    border: 1px solid ${({ $variant }) => checkboxVariants[$variant].borderColor};
    border-radius: 0.2rem;
    background-color: ${colors.palette.white};
    position: relative;

    &::before {
      content: "";
      position: absolute;
      width: 100%;
      height: 100%;
      background-color: ${({ $variant }) => checkboxVariants[$variant].backgroundColor};
      background-image: url(${({ $variant }) => checkboxVariants[$variant].checkmark});
      background-position: center;
      background-repeat: no-repeat;
      background-size: 80%;
      transform: scale(0);
      transition: transform 0.2s ease-in-out;
      border-radius: 0.2rem;
    }

    &:checked::before {
      transform: scale(1);
    }

    &:disabled {
      opacity: 0.4;
      cursor: not-allowed;
    }
  }
`;

export interface CheckboxProps extends PropsWithChildren<Omit<RACCheckboxProps, 'children'>> {
  size?: Size;
  variant?: Variant;
  bold?: boolean;
}

export const Checkbox = ({
  size = 1.6,
  variant = 'primary',
  bold = false,
  children,
  ...props
}: CheckboxProps) => {
  return (
    <StyledCheckbox $size={size} $variant={variant} $bold={bold} {...props}>
      {children}
    </StyledCheckbox>
  );
};