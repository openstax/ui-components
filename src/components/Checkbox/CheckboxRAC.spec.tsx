import { CheckboxRAC } from './CheckboxRAC';
import renderer from 'react-test-renderer';

describe('CheckboxRAC', () => {
  it('matches snapshot', () => {
    const tree = renderer.create(
      <CheckboxRAC>Click Me</CheckboxRAC>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('handles options', () => {
    const tree = renderer.create(
      <CheckboxRAC bold size={2} variant='light'>Click Me</CheckboxRAC>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('handles disabled state', () => {
    const tree = renderer.create(
      <CheckboxRAC isDisabled>Click Me</CheckboxRAC>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('supports slot="selection"', () => {
    const tree = renderer.create(
      <CheckboxRAC slot="selection">Click Me</CheckboxRAC>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});