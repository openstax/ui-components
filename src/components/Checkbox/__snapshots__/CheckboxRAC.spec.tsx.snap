// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckboxRAC handles disabled state 1`] = `
<label
  className="sc-bczRLJ kDipcy"
  data-disabled={true}
  data-rac=""
  data-react-aria-pressable={true}
  onClick={[Function]}
  onDragStart={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchCancel={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
>
  <span
    style={
      Object {
        "border": 0,
        "clip": "rect(0 0 0 0)",
        "clipPath": "inset(50%)",
        "height": "1px",
        "margin": "-1px",
        "overflow": "hidden",
        "padding": 0,
        "position": "absolute",
        "whiteSpace": "nowrap",
        "width": "1px",
      }
    }
  >
    <input
      checked={false}
      data-react-aria-pressable={true}
      disabled={true}
      onBlur={[Function]}
      onChange={[Function]}
      onClick={[Function]}
      onDragStart={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchCancel={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      type="checkbox"
    />
  </span>
  <div
    data-slot="selection"
    slot="selection"
  />
  <span
    className="react-aria-Text"
  >
    Click Me
  </span>
</label>
`;

exports[`CheckboxRAC handles options 1`] = `
<label
  className="sc-bczRLJ kJYYoi"
  data-rac=""
  data-react-aria-pressable={true}
  onClick={[Function]}
  onDragStart={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchCancel={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
>
  <span
    style={
      Object {
        "border": 0,
        "clip": "rect(0 0 0 0)",
        "clipPath": "inset(50%)",
        "height": "1px",
        "margin": "-1px",
        "overflow": "hidden",
        "padding": 0,
        "position": "absolute",
        "whiteSpace": "nowrap",
        "width": "1px",
      }
    }
  >
    <input
      checked={false}
      data-react-aria-pressable={true}
      disabled={false}
      onBlur={[Function]}
      onChange={[Function]}
      onClick={[Function]}
      onDragStart={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchCancel={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="checkbox"
    />
  </span>
  <div
    data-slot="selection"
    slot="selection"
  />
  <span
    className="react-aria-Text"
  >
    Click Me
  </span>
</label>
`;

exports[`CheckboxRAC matches snapshot 1`] = `
<label
  className="sc-bczRLJ leFkQj"
  data-rac=""
  data-react-aria-pressable={true}
  onClick={[Function]}
  onDragStart={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchCancel={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
>
  <span
    style={
      Object {
        "border": 0,
        "clip": "rect(0 0 0 0)",
        "clipPath": "inset(50%)",
        "height": "1px",
        "margin": "-1px",
        "overflow": "hidden",
        "padding": 0,
        "position": "absolute",
        "whiteSpace": "nowrap",
        "width": "1px",
      }
    }
  >
    <input
      checked={false}
      data-react-aria-pressable={true}
      disabled={false}
      onBlur={[Function]}
      onChange={[Function]}
      onClick={[Function]}
      onDragStart={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchCancel={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="checkbox"
    />
  </span>
  <div
    data-slot="selection"
    slot="selection"
  />
  <span
    className="react-aria-Text"
  >
    Click Me
  </span>
</label>
`;

exports[`CheckboxRAC supports slot="selection" 1`] = `
<label
  className="sc-bczRLJ leFkQj"
  data-rac=""
  data-react-aria-pressable={true}
  onClick={[Function]}
  onDragStart={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchCancel={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  slot="selection"
>
  <span
    style={
      Object {
        "border": 0,
        "clip": "rect(0 0 0 0)",
        "clipPath": "inset(50%)",
        "height": "1px",
        "margin": "-1px",
        "overflow": "hidden",
        "padding": 0,
        "position": "absolute",
        "whiteSpace": "nowrap",
        "width": "1px",
      }
    }
  >
    <input
      checked={false}
      data-react-aria-pressable={true}
      disabled={false}
      onBlur={[Function]}
      onChange={[Function]}
      onClick={[Function]}
      onDragStart={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchCancel={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="checkbox"
    />
  </span>
  <div
    data-slot="selection"
    slot="selection"
  />
  <span
    className="react-aria-Text"
  >
    Click Me
  </span>
</label>
`;
